README.rst
====================
django-nose-selenium
====================

Adds selenium testing support to your nose test suite.

To use, run nosetests with the ``--with-selenium`` flag.

-----------------
Why use Selenium?
-----------------

Selenium is a portable testing framework for web applications. It allows you to
write tests that run in the browser to test your user interface and javascript
code that is not available through the usual testing channels. See the examples
below to get a clearer impression of what selenium tests can provide.

django-nose-selenium allows you to write and run selenium tests the same way as
usual django unit tests.

------------
Requirements
------------

The plugin expects that you have configured the django-nose_ app. In a nutshell,
this is done by running ``pip install django-nose``, adding ``django_nose`` to
``INSTALLED_APPS`` and setting ``TEST_RUNNER`` to
``django_nose.NoseTestSuiteRunner`` in the settings.py.

.. _django-nose: https://github.com/jbalogh/django-nose

------------
Installation
------------

From PyPI::

   pip install django-nose-selenium

`In-development version
<https://github.com/weluse/django-nose-selenium/tarball/master#egg=django-nose-selenium-dev>`_
via Pip::

   pip install django-nose-selenium==dev

Directly from Git::

   pip install -e
   git://github.com/weluse/django-nose-selenium.git#egg=django-nose-selenium

---------------
Django Settings
---------------

.. _base_settings:

The plugin supports the following settings:

   * SELENIUM_HOST, default: `127.0.0.1`
   * SELENIUM_PORT, default: `4444`
   * SELENIUM_BROWSER_COMMAND, default: `chrome`
   * SELENIUM_URL_ROOT, default: `http://127.0.0.1:8000`
   * FORCE_SELENIUM_TESTS, default: `False`. By default, SocketErrors cause the
     tests to be skipped. This options causes the tests to fail when the
     Selenium server is unavailable.

-----
Usage
-----

Define the class variable ``selenium_test = True`` in your nose test class.
You can use ``self.selenium`` to access a selenium instance with the given
options::


   class TestSelenium(TestCase):

       selenium_test = True

       def test_start(self):
           """Tests the start page."""

           self.selenium.open("/")

To run this test, you have to pass the option ``--with-selenium`` to the Django
management command test::

   python manage.py test --with-selenium

Alternatively, django-nose-selenium provides a mixin that has the benefit of
raising a SkipTest exception if the plugin was not loaded and the selenium
attribute is accessed::


   from noseselenium.cases import SeleniumTestCaseMixin


   class TestSelenium(TestCase, SeleniumTestCaseMixin):

       def test_start(self):
           """Tests the start page."""

           self.selenium.open("/")

Fixtures
--------

The default fixtures of django are run in transactions and not available to a
live testing server, therefore `noseselenium` provides an option to load **and
commit** fixtures to the database automatically. Please note that there's no
automatic rollback, so the data will stay in your test database for the rest of
the run if you don't provide a custom teardown strategy.

::

   from noseselenium.cases import SeleniumTestCaseMixin


   class TestUserLogin(TestCase, SeleniumTestCaseMixin):

       selenium_fixtures = ['user_data.json']

       def tearDown(self):
           # Remove data from user_data.json

       def test_login(self):
           """Tests the login page."""

           sel = self.selenium
           sel.open("/login/")
           sel.type("id_username", "pascal")
           sel.type("id_password", "iwantapony")
           sel.click("//form[@id='myform']/p/button")
           sel.wait_for_page_to_load(5000)
           self.failUnless(self.is_text_present("Hello, Pascal!"))

To enable selenium fixtures, nosetests must be called with the
additional ``--with-selenium-fixtures`` flag.


Liveserver
----------

`noseselenium` provides expiremental support for running a live server that
Selenium can connect to. Currently, there's a threaded server that reuses
django's development webserver and a cherrypy implementation. It's recommended
you use the cherrypy one as the django devserver is certainly not designed to
run in a multi-threaded environment.

The liveserver plugin introduces two new configuration options:

   * LIVE_SERVER_ADDRESS, defaults to `0.0.0.0`
   * LIVE_SERVER_PORT, defaults to `8080`
   * LIVE_SERVER_STATIC, boolean that defaults to True. If enabled, the live
     server enables serving of static files via the
     ``django.contrib.staticfiles`` app.

These should match your `Selenium Settings`__.

__ base_settings_

To start the liveserver, nosetest is called with either the
``--with-djangoliveserver`` or preferably the ``--with-cherrypyliveserver``
flag.


CHANGES
django-nose-selenium Changelog
==============================

Here you can see the list of changes between the releases.

Version 0.7.3
-------------

Bugfix release, released on 2011-08-15

- Fixed regression regarding static files serving with the CherryPy liveserver
  plugin.

Version 0.7.2
-------------

Bugfix release, released on 2011-08-03

- Restored compatibility with Django 1.2 (#7, reported by Issac Kelly)


example/myapp/tests.py
from django.test import TestCase
from noseselenium.cases import SeleniumTestCaseMixin
from selenium import webdriver


class ExampleWebdriverTestCase(TestCase, SeleniumTestCaseMixin):

    def get_driver(self):
        """Overrides the default driver to set project-specific settings."""

        return webdriver.Chrome()


class View1WebdriverTestCase(ExampleWebdriverTestCase):
    def test_view(self):
        driver = self.selenium.driver
        self.selenium.open('/view1')
        self.assertEquals('Hello, World!', driver.title)


example/settings.py
import os


DEBUG = True
TEMPLATE_DEBUG = DEBUG

ADMINS = (
    # ('Your Name', 'your_email@example.com'),
)

MANAGERS = ADMINS

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3', # Add 'postgresql_psycopg2', 'postgresql', 'mysql', 'sqlite3' or 'oracle'.
        'NAME': 'example.db',                      # Or path to database file if using sqlite3.
        'TEST_NAME': 'example_test.db',
        'USER': '',                      # Not used with sqlite3.
        'PASSWORD': '',                  # Not used with sqlite3.
        'HOST': '',                      # Set to empty string for localhost. Not used with sqlite3.
        'PORT': '',                      # Set to empty string for default. Not used with sqlite3.
    }
}

PROJECT_ROOT = os.path.abspath(os.path.dirname(__file__))

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# On Unix systems, a value of None will cause Django to use the same
# timezone as the operating system.
# If running in a Windows environment this must be set to the same as your
# system time zone.
TIME_ZONE = 'America/Chicago'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'en-us'

SITE_ID = 1

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# If you set this to False, Django will not format dates, numbers and
# calendars according to the current locale
USE_L10N = True

# Absolute filesystem path to the directory that will hold user-uploaded files.
# Example: "/home/media/media.lawrence.com/media/"
MEDIA_ROOT = ''

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash.
# Examples: "http://media.lawrence.com/media/", "http://example.com/media/"
MEDIA_URL = ''

# Absolute path to the directory static files should be collected to.
# Don't put anything in this directory yourself; store your static files
# in apps' "static/" subdirectories and in STATICFILES_DIRS.
# Example: "/home/media/media.lawrence.com/static/"
STATIC_ROOT = ''

# URL prefix for static files.
# Example: "http://media.lawrence.com/static/"
STATIC_URL = '/static/'

# URL prefix for admin static files -- CSS, JavaScript and images.
# Make sure to use a trailing slash.
# Examples: "http://foo.com/static/admin/", "/static/admin/".
ADMIN_MEDIA_PREFIX = '/static/admin/'

# Additional locations of static files
STATICFILES_DIRS = (
    # Put strings here, like "/home/html/static" or "C:/www/django/static".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
)

# List of finder classes that know how to find static files in
# various locations.
STATICFILES_FINDERS = (
    'django.contrib.staticfiles.finders.FileSystemFinder',
    'django.contrib.staticfiles.finders.AppDirectoriesFinder',
#    'django.contrib.staticfiles.finders.DefaultStorageFinder',
)

# Make this unique, and don't share it with anybody.
SECRET_KEY = '7$_ntaffut@e^ho@dwl^ety#u*0(rt!@f&0qx7j$szhay@#1ja'

# List of callables that know how to import templates from various sources.
TEMPLATE_LOADERS = (
    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
#     'django.template.loaders.eggs.Loader',
)

MIDDLEWARE_CLASSES = (
    'django.middleware.common.CommonMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
)

ROOT_URLCONF = 'example.urls'

TEMPLATE_DIRS = (
    # Put strings here, like "/home/html/django_templates" or "C:/www/django/templates".
    # Always use forward slashes, even on Windows.
    # Don't forget to use absolute paths, not relative paths.
)

INSTALLED_APPS = (
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.messages',
    'django.contrib.staticfiles',
    # Uncomment the next line to enable the admin:
    # 'django.contrib.admin',
    # Uncomment the next line to enable admin documentation:
    # 'django.contrib.admindocs',
    'example.myapp',
)

# A sample logging configuration. The only tangible logging
# performed by this configuration is to send an email to
# the site admins on every HTTP 500 error.
# See http://docs.djangoproject.com/en/dev/topics/logging for
# more details on how to customize your logging configuration.
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False,
    'handlers': {
        'mail_admins': {
            'level': 'ERROR',
            'class': 'django.utils.log.AdminEmailHandler'
        }
    },
    'loggers': {
        'django.request': {
            'handlers': ['mail_admins'],
            'level': 'ERROR',
            'propagate': True,
        },
    }
}

import sys

# Modify the path to allow the uninstalled use of the noseselenium plugin.
sys.path.insert(0, os.path.join(PROJECT_ROOT, os.path.pardir))

TEST_RUNNER = 'django_nose.NoseTestSuiteRunner'
NOSE_PLUGINS = [
    'noseselenium.plugins.SeleniumPlugin',
    'noseselenium.plugins.DjangoLiveServerPlugin',
]
NOSE_ARGS = ["--with-selenium", '--with-djangoliveserver']


noseselenium/cases.py
# -*- coding: utf-8 -*-
"""
noseselenium.cases
~~~~~~~~~~~~~~~~~~

Testcases for selenium.

:copyright: 2010, Pascal Hartig <phartig@weluse.de>
:license: BSD, see LICENSE for more details.
"""

from nose.plugins.skip import SkipTest
from selenium import webdriver


class SeleniumSkipper(object):
    """
    A class that raises a SkipTest exception when an attribute is
    accessed.
    """

    def __getattr__(self, name):
        raise SkipTest("SeleniumPlugin not enabled.")


class SeleniumTestCaseMixin(object):
    """
    Apply this to your TestCase base class to get Webdriver functionality in
    your tests. If the selenium attribute is accessed without having
    :class:`SeleniumPlugin` enabled, the test is skipped.
    """

    selenium = SeleniumSkipper()
    selenium_test = True
    start_live_server = True

    def get_driver(self):
        """Builds the driver instance and returns it. Override this to use
        custom options.
        """

        return webdriver.Firefox()


noseselenium/plugins.py
# -*- coding: utf-8 -*-
"""
noseselenium.plugins
~~~~~~~~~~~~~~~~~~~~

Provides the nose plugins.
Heavily inspired by `django-sane-testing`_.

_django-sane-testing: http://github.com/Almad/django-sane-testing/

:copyright: 2010-2011, Pascal Hartig <phartig@weluse.de>
:license: BSD, see LICENSE for more details.
"""

import socket
import nose
import time
import threading
import django
import selenium

from nose.plugins import Plugin
from django.conf import settings
from selenium.webdriver.emulation.selenium1 import DrivenSelenium
from unittest import TestCase
# Liveserver imports
from SocketServer import ThreadingMixIn
from BaseHTTPServer import HTTPServer
from django.core.handlers.wsgi import WSGIHandler
from django.core.servers.basehttp import WSGIRequestHandler, \
        AdminMediaHandler, WSGIServerException
from django.db.backends.creation import TEST_DATABASE_PREFIX


def _get_test_db_name(connection):
    """Tries to build the test database name like django does."""

    if connection.settings_dict['TEST_NAME']:
        return connection.settings_dict['TEST_NAME']
    else:
        old_name = connection.settings_dict['NAME']
        if old_name.startswith(TEST_DATABASE_PREFIX):
            return old_name
        else:
            return TEST_DATABASE_PREFIX + \
                    old_name


def _set_autocommit(connection):
        """Make sure a connection is in autocommit mode."""
        if hasattr(connection.connection, "autocommit"):
            if callable(connection.connection.autocommit):
                connection.connection.autocommit(True)
            else:
                connection.connection.autocommit = True
        elif hasattr(connection.connection, "set_isolation_level"):
            connection.connection.set_isolation_level(0)


def _setup_test_db():
    """Activates a test dbs without recreating them."""

    from django.db import connections

    for alias in connections:
        connection = connections[alias]
        connection.close()

        test_db_name = _get_test_db_name(connection)
        connection.settings_dict['NAME'] = test_db_name

        # SUPPORTS_TRANSACTIONS is not needed in newer versions of djangoo
        if not hasattr(connection.features, 'supports_transactions'):
            can_rollback = connection.creation._rollback_works()
            connection.settings_dict['SUPPORTS_TRANSACTIONS'] = can_rollback

        # Trigger side effects.
        connection.cursor()
        _set_autocommit(connection)


def _patch_static_handler(handler):
    """Patch in support for static files serving if supported and enabled.
    """

    if django.VERSION[:2] < (1, 3):
        return

    from django.contrib.staticfiles.handlers import StaticFilesHandler
    return StaticFilesHandler(handler)


def get_test_case_class(nose_test):
    """
    Extracts the class from the nose tests that depends on whether it's a
    method test case or a function test case.
    """
    if isinstance(nose_test.test, nose.case.MethodTestCase):
        return nose_test.test.test.im_class
    else:
        return nose_test.test.__class__


class SeleniumPlugin(Plugin):
    """
    Adds a selenium attribute to the nose test case and reads the parameters
    from the django config.
    Only works with class based tests, so far.
    """

    activation_parameter = "--with-selenium"
    name = "selenium"
    score = 80

    def startTest(self, test):
        """
        When preparing the test, inject a selenium instance.
        """

        test_case = get_test_case_class(test)
        if getattr(test_case, "selenium_test", False):
            self._inject_selenium(test)

    def stopTest(self, test):
        """
        Destroys the selenium connection.
        """

        test_case = get_test_case_class(test)
        if getattr(test_case, 'selenium_started', False):
            if isinstance(test.test, nose.case.MethodTestCase):
                im_self = test.test.test.im_self
            elif isinstance(test.test, TestCase):
                im_self = test.test.run.im_self

            im_self.selenium.driver.quit()

    def _inject_selenium(self, test):
        """
        Injects a selenium instance into the method.
        """
        test_case = get_test_case_class(test)
        test_case.selenium_plugin_started = True

        if not hasattr(test.test, 'get_driver'):
            raise RuntimeError("Test case does not implement a get_driver() "
                               "method.")

        driver = test.test.get_driver()
        url_root = getattr(settings, 'SELENIUM_URL_ROOT',
                           "http://127.0.0.1:8080")
        sel = DrivenSelenium(driver, url_root)

        test.test.selenium = sel
        test_case.selenium_started = True


class SeleniumFixturesPlugin(Plugin):
    """
    Loads fixtures defined in the attribute `selenium_fixtures`. It does,
    however, not take care of removing them after the test or even the whole
    test case is run.
    Django fixtures are usually run in transactions so a test server accessing
    the test database won't be able access the data.
    """

    activation_parameter = "--with-selenium-fixtures"
    name = "selenium-fixtures"
    score = 80

    def startTest(self, test):
        """
        When preparing the database, check for the `selenium_fixtures`
        attribute and load those.
        """

        from django.test.testcases import call_command

        test_case = get_test_case_class(test)
        fixtures = getattr(test_case, "selenium_fixtures", [])

        if fixtures:
            call_command('loaddata', *fixtures, **{
                'verbosity': 1,
                # Necessary to let the test server access them.
                'commit': True
            })


class StoppableWSGIServer(ThreadingMixIn, HTTPServer):
    """WSGIServer with short timeout, so that server thread can stop this
    server.

    This implementation is again from django-sane-testing, while the original
    code is taken from django ticket #2879 which proposes a live server in the
    django core.
    """

    application = None

    def __init__(self, server_address, RequestHandlerClass=None):
        HTTPServer.__init__(self, server_address, RequestHandlerClass)

    def server_bind(self):
        """Bind server to socket. Overrided to store server name and
        set timeout.
        """

        try:
            HTTPServer.server_bind(self)
        except Exception as e:
            raise WSGIServerException(e)

        self.setup_environ()
        self.socket.settimeout(1)

    def get_request(self):
        """Checks for timeout when getting request."""
        sock, address = self.socket.accept()
        return (sock, address)

    def setup_environ(self):
        """Set up a basic environment."""

        self.base_environ = {
            'SERVER_NAME': self.server_name,
            'GATEWAY_INTERFACE': 'CGI/1.1',
            'SERVER_PORT': str(self.server_port),
            'REMOTE_HOST': '',
            'CONTENT_LENGTH': '',
            'SCRIPT_NAME': '',
        }

    def get_app(self):
        return self.application

    def set_app(self,application):
        self.application = application


class AbstractLiveServerPlugin(Plugin):
    """Base class for live servers."""

    score = 70

    def __init__(self):
        Plugin.__init__(self)
        self.server_started = False
        self.server_thread = None

    def start_server(self):
        raise NotImplementedError()

    def stop_server(self):
        raise NotImplementedError()

    def startTest(self, test):
        """Starts the live server."""

        from django.conf import settings

        test_case = get_test_case_class(test)

        if not self.server_started and \
           getattr(test_case, "start_live_server", False):

            _setup_test_db()

            # Raises an exception if not.
            settings.TEST_MODE = True

            self.start_server(
                address=getattr(settings, 'LIVE_SERVER_ADDRESS',
                                '0.0.0.0'),
                port=getattr(settings, 'LIVE_SERVER_PORT',
                             8080),
                serve_static=getattr(settings, 'LIVE_SERVER_STATIC', True)
            )

            self.server_started = True
            setattr(test_case, 'http_plugin_started', True)

    def stopTest(self, test):
        """Stops the live server if necessary."""

        test_case = get_test_case_class(test)
        if self.server_started and \
           getattr(test_case, 'http_plugin_started', False):

            self.stop_server()
            self.server_started = False


class TestServerThread(threading.Thread):
    """Thread for running a http server while tests are running."""

    def __init__(self, address, port, serve_static=True):
        self.address = address
        self.port = port
        self.serve_static = serve_static
        self._stopevent = threading.Event()
        self.started = threading.Event()
        self.error = None
        super(TestServerThread, self).__init__()

    def run(self):
        """Sets up test server and loops over handling http requests."""
        try:
            handler = AdminMediaHandler(WSGIHandler())
            if self.serve_static:
                handler = _patch_static_handler(handler)

            server_address = (self.address, self.port)
            httpd = StoppableWSGIServer(server_address, WSGIRequestHandler)
            httpd.application = handler
            self.started.set()
        except WSGIServerException as err:
            self.error = err
            self.started.set()
            return

        # Loop until we get a stop event.
        while not self._stopevent.isSet():
            httpd.handle_request()

    def join(self, timeout=None):
        """Stop the thread and wait for it to finish."""
        self._stopevent.set()
        threading.Thread.join(self, timeout)


class DjangoLiveServerPlugin(AbstractLiveServerPlugin):
    """
    Patch Django on fly and start live HTTP server, if TestCase is inherited
    from HttpTestCase or start_live_server attribute is set to True.

    Taken from Michael Rogers implementation from `getwindmill.com
    <http://trac.getwindmill.com/browser/trunk/windmill/authoring/
    djangotest.py>`_.
    """
    name = 'djangoliveserver'
    activation_parameter = '--with-djangoliveserver'

    def start_server(self, address='0.0.0.0', port=8000, serve_static=True):
        self.server_thread = TestServerThread(address, port, serve_static)
        self.server_thread.start()
        self.server_thread.started.wait()
        if self.server_thread.error:
            raise self.server_thread.error

    def stop_server(self):
        self.server_thread.join()


class CherryPyLiveServerPlugin(AbstractLiveServerPlugin):
    """Live server plugin using cherrypy instead of the django server,
    that got its issues. Original code by Mikeal Rogers, released under
    Apache License.
    """

    name = 'cherrypyliveserver'
    activation_parameter = '--with-cherrypyliveserver'

    def start_server(self, address='0.0.0.0', port=8000, serve_static=True):
        from cherrypy.wsgiserver import CherryPyWSGIServer
        from threading import Thread

        _application = AdminMediaHandler(WSGIHandler())
        if serve_static:
            _application = _patch_static_handler(_application)

        def application(environ, start_response):
            environ['PATH_INFO'] = environ['SCRIPT_NAME'] + \
                    environ['PATH_INFO']

            return _application(environ, start_response)

        self.httpd = CherryPyWSGIServer((address, port), application,
                                        server_name='django-test-http')
        self.httpd_thread = Thread(target=self.httpd.start)
        self.httpd_thread.start()
        # FIXME: This could be avoided by passing self to thread class starting
        # django and waiting for Event lock
        time.sleep(.5)

    def stop_server(self):
        self.httpd.stop()
